### PR TITLE
Azure: Configuration for user identity authentication in datasources (Experimental)

### PR DESCRIPTION
**Pull request by kostrse**
**What this PR does / why we need it**:

This PR introduces additional configuration settings for Azure datasources which support user identity authentication on behalf of currently signed-in Grafana user.

These settings include a logical toggle `user_identity_enabled` to explicitly enable the user identity authentication method, as well as token endpoint settings for exchange of user's token to access token for the resource endpoint .

Typical configuration:

```ini
[auth.azuread]
enabled = true

auth_url = https://login.microsoftonline.com/fd719c11-a91c-40fd-8379-1e6cd3c59568//oauth2/v2.0/authorize
token_url = https://login.microsoftonline.com/fd719c11-a91c-40fd-8379-1e6cd3c59568/oauth2/v2.0/token
client_id = f85aa887-490d-4fac-9306-9b99ad0aa31d
client_secret = 87808761-ff7b-492e-bb0d-5de2437ffa55

[azure]
user_identity_enabled = true
```

In the example above the some AAD application will be used for Grafana sign-in and datasource token requests.

On-Behalf-Of (different AAD apps for Grafana sign-in and datasource token requests):

```ini
[auth.azuread]
enabled = true

auth_url = https://login.microsoftonline.com/fd719c11-a91c-40fd-8379-1e6cd3c59568//oauth2/v2.0/authorize
token_url = https://login.microsoftonline.com/fd719c11-a91c-40fd-8379-1e6cd3c59568/oauth2/v2.0/token
client_id = f85aa887-490d-4fac-9306-9b99ad0aa31d
client_secret = 87808761-ff7b-492e-bb0d-5de2437ffa55

[azure]
user_identity_enabled = true

# Use different AAD app (OBO)
user_identity_client_id = 4fc34037-97bd-4e84-9db4-86238c78e32a
user_identity_client_secret = 4479f5a6-444c-4271-8790-60eeb42225ae
```

Advanced scenario, configure custom token endpoint:

```ini
[azure]
user_identity_enabled = true

user_identity_token_url = https://custom-token-endpoint/oauth2/v2.0/token
user_identity_client_id = 4fc34037-97bd-4e84-9db4-86238c78e32a
user_identity_client_secret = 4479f5a6-444c-4271-8790-60eeb42225ae
```

**Which issue(s) this PR fixes**:

Related https://github.com/grafana/grafana-azure-sdk-go/issues/6

**Special notes for your reviewer**:

Grafana host itself doesn't use these settings, they are intended to be passed to external Azure plugins like Azure Data Explorer.

### CODE REVIEW

**Comment by kostrse** [comment]
> From initial testing on this it seems there is a bug and the `cfg.Azure.UserIdentityEnabled` needs to be passed into `token_provider_azure.go`.

`token_provider_azure.go` from `pluginproxy` was initially used by Azure Monitor when Azure Monitor had declaratively defined routes in `plugin.json` but it has changed long time ago. I don't know of any datasource which actually uses it.

So, I'm not planning to add support for user identity auth into `token_provider_azure.go`.

I'm changing signature of the `aztokenprovider.NewAzureAccessTokenProvider` ([this](https://github.com/kostrse/grafana/blob/ae288e27ebe2967210c416100a51ecd622c63a92/pkg/api/pluginproxy/token_provider_azure.go#L23) line) in Azure SDK, I will updated it when I will be integrating updated version of SDK (not this PR).

If you're trying to make it work end to end localy, just add third argument `false` there.

```go
	tokenProvider, err := aztokenprovider.NewAzureAccessTokenProvider(cfg.Azure, credentials, false)
```

File: conf/defaults.ini [hunk 1]
```diff
@@ -804,6 +804,23 @@ managed_identity_enabled = false
 # Should be set for user-assigned identity and should be empty for system-assigned identity
 managed_identity_client_id =
 
+# Specifies whether user identity authentication on behalf of currently signed-in user should be enabled in datasources
+# which support it (may require AAD authentication enabled)
+# Disabled by default, needs to be explicitly enabled
+user_identity_enabled = false
+
+# Override token URL for given Azure Active Directory
```
Comment (aangelisc) [inline conf/defaults.ini:812]:
```suggestion
# Override token URL for Azure Active Directory
```

File: conf/defaults.ini [hunk 2]
```diff
@@ -804,6 +804,23 @@ managed_identity_enabled = false
...
+# By default is the same as token URL configured for AAD authentication settings
+user_identity_token_url =
+
+# Override ADD application ID which would be used to exchange user's token to an access token to the datasource
```
Comment (aangelisc) [inline conf/defaults.ini:816]:
```suggestion
# Override ADD application ID which would be used to exchange users token to an access token for the datasource
```

File: conf/sample.ini [hunk 3]
```diff
@@ -776,6 +776,23 @@
...
+# Specifies whether user identity authentication on behalf of currently signed-in user should be enabled in datasources
+# which support it (may require AAD authentication enabled)
+# Disabled by default, needs to be explicitly enabled
+;user_identity_enabled = false
+
+# Override token URL for given Azure Active Directory
+# By default is the same as token URL configured for AAD authentication settings
+;user_identity_token_url =
+
+# Override ADD application ID which would be used to exchange user's token to an access token to the datasource
```
Comment (aangelisc) [inline conf/sample.ini:788]:
```suggestion
# Override ADD application ID which would be used to exchange users token to an access token for the datasource
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 4]
```diff
@@ -1117,6 +1117,30 @@ The client ID to use for user-assigned managed identity.
 
 Should be set for user-assigned identity and should be empty for system-assigned identity.
 
+### user_identity_enabled
+
+Specifies whether user identity authentication on behalf of currently signed-in user should be enabled in datasources which support it (may require AAD authentication enabled).
```
Comment (aangelisc) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1122]: Same as above

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 5]
```diff
@@ -1117,6 +1117,30 @@ The client ID to use for user-assigned managed identity.
...
+
+### user_identity_token_url
+
+Override token URL for given Azure Active Directory.
+
+By default is the same as token URL configured for AAD authentication settings.
+
+### user_identity_client_id
+
+Override ADD application ID which would be used to exchange user's token to an access token to the datasource.
```
Comment (aangelisc) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1134]:
```suggestion
Override ADD application ID which would be used to exchange users token to an access token for the datasource.
```

**Review by aangelisc — APPROVED** [review]
Just a couple nits/comments 😊 You also have a failing lint check due to:
```
pkg/setting/setting_azure_test.go:67: unnecessary leading newline (whitespace)
	t.Run("User Identity", func(t *testing.T) {
```

**Review by marefr — APPROVED** [review]
LGTM

**Pull request merged**